File: openfeature-provider/go/confidence/state_fetcher.go
package confidence

import (
	"context"
	"crypto/sha256"
	"encoding/hex"
	"fmt"
	"io"
	"log/slog"
	"net/http"
	"sync/atomic"
	"time"

	pb "github.com/spotify/confidence-resolver/openfeature-provider/go/confidence/proto"
	adminv1 "github.com/spotify/confidence-resolver/openfeature-provider/go/confidence/proto/confidence/flags/admin/v1"
	"google.golang.org/protobuf/proto"
)

// StateProvider is an interface for providing resolver state and account ID
type StateProvider interface {
	Provide(ctx context.Context) ([]byte, string, error)
}

// FlagsAdminStateFetcher fetches and updates the resolver state from the CDN
type FlagsAdminStateFetcher struct {
	clientSecret     string
	etag             atomic.Value // stores string
	rawResolverState atomic.Value // stores []byte
	accountID        atomic.Value // stores string
	HTTPClient       *http.Client // Exported for testing
	logger           *slog.Logger
}

// Compile-time interface conformance check
var _ StateProvider = (*FlagsAdminStateFetcher)(nil)

// NewFlagsAdminStateFetcher creates a new FlagsAdminStateFetcher
func NewFlagsAdminStateFetcher(
	clientSecret string,
	logger *slog.Logger,
) *FlagsAdminStateFetcher {
	f := &FlagsAdminStateFetcher{
		clientSecret: clientSecret,
		logger:       logger,
		HTTPClient: &http.Client{
			Timeout: 30 * time.Second,
		},
	}
	// Initialize with empty state
	emptyState := &adminv1.ResolverState{}
	if b, err := proto.Marshal(emptyState); err == nil {
		f.rawResolverState.Store(b)
	}
	return f
}

// GetRawState returns the current raw resolver state
func (f *FlagsAdminStateFetcher) GetRawState() []byte {
	if state := f.rawResolverState.Load(); state != nil {
		return state.([]byte)
	}
	return nil
}

// GetAccountID returns the account ID
func (f *FlagsAdminStateFetcher) GetAccountID() string {
	if accountID := f.accountID.Load(); accountID != nil {
		return accountID.(string)
	}
	return ""
}

// Reload fetches and updates the state if it has changed
func (f *FlagsAdminStateFetcher) Reload(ctx context.Context) error {
	if err := f.fetchAndUpdateStateIfChanged(ctx); err != nil {
		f.logger.Warn("Failed to reload, ignoring reload", "error", err)
		return err
	}
	return nil
}

// Provide implements the StateProvider interface
// Returns the latest resolver state and account ID, fetching it if needed
// On error, returns cached state (if available) to maintain availability
func (f *FlagsAdminStateFetcher) Provide(ctx context.Context) ([]byte, string, error) {
	// Try to fetch the latest state
	err := f.Reload(ctx)
	// Always return the current state and accountID (cached or fresh)
	// This ensures availability even if fetch fails
	return f.GetRawState(), f.GetAccountID(), err
}

// fetchAndUpdateStateIfChanged fetches the state from the CDN if it has changed
func (f *FlagsAdminStateFetcher) fetchAndUpdateStateIfChanged(ctx context.Context) error {
	// Build CDN URL using SHA256 hash of client secret
	hash := sha256.Sum256([]byte(f.clientSecret))
	hashHex := hex.EncodeToString(hash[:])
	cdnURL := "https://confidence-resolver-state-cdn.spotifycdn.com/" + hashHex

	req, err := http.NewRequestWithContext(ctx, http.MethodGet, cdnURL, nil)
	if err != nil {
		return err
	}

	// Add If-None-Match header if we have a previous ETag
	if previousEtag := f.etag.Load(); previousEtag != nil {
		req.Header.Set("If-None-Match", previousEtag.(string))
	}

	resp, err := f.HTTPClient.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	// Check if content was modified
	if resp.StatusCode == http.StatusNotModified {
		// Not modified, nothing to update
		return nil
	}

	if resp.StatusCode != http.StatusOK {
		return fmt.Errorf("unexpected status code: %d", resp.StatusCode)
	}

	// Read the new state
	bytes, err := io.ReadAll(resp.Body)
	if err != nil {
		return err
	}

	// Parse SetResolverStateRequest
	stateRequest := &pb.SetResolverStateRequest{}
	if err := proto.Unmarshal(bytes, stateRequest); err != nil {
		return fmt.Errorf("failed to unmarshal SetResolverStateRequest: %w", err)
	}

	// Extract account ID and state bytes
	f.accountID.Store(stateRequest.AccountId)

	// Get and store the new ETag
	etag := resp.Header.Get("ETag")
	f.etag.Store(etag)

	// Update the raw state (state is already in bytes format)
	f.rawResolverState.Store(stateRequest.State)

	f.logger.Info("Loaded resolver state", "etag", etag, "account", stateRequest.AccountId)

	return nil
}

File: openfeature-provider/python/confidence/state_fetcher.py
"""State fetcher for downloading flag resolver state from CDN."""

import logging
from typing import Optional

import httpx

logger = logging.getLogger(__name__)


class StateFetcher:
    """Fetches and caches resolver state from the Confidence CDN."""

    def __init__(self, client_secret: str, timeout: float = 30.0):
        """Initialize the state fetcher.

        Args:
            client_secret: Client secret for authentication
            timeout: HTTP request timeout in seconds
        """
        self.client_secret = client_secret
        self.cdn_url = f"https://confidence-resolver-state-cdn.spotifycdn.com/{client_secret}"
        self.timeout = timeout

        # State caching
        self._etag: Optional[str] = None
        self._cached_state: Optional[bytes] = None

        # HTTP client
        self._http_client = httpx.AsyncClient(timeout=timeout)

    async def close(self) -> None:
        """Close the HTTP client."""
        await self._http_client.aclose()

    async def fetch_state(self) -> bytes:
        """Fetch the resolver state from CDN.

        The CDN returns raw bytes (typically a serialized protobuf).

        Returns:
            Raw state bytes from CDN

        Raises:
            httpx.HTTPError: If the HTTP request fails
        """
        headers = {}
        if self._etag:
            headers["If-None-Match"] = self._etag

        try:
            response = await self._http_client.get(self.cdn_url, headers=headers)

            # Check if content was modified
            if response.status_code == 304:
                logger.debug("State not modified (304), using cached state")
                # Return cached state if available
                if self._cached_state is not None:
                    return self._cached_state
                # If no cache, this is unexpected but try again without ETag
                logger.warning("Received 304 but no cached state, retrying without ETag")
                response = await self._http_client.get(self.cdn_url)

            response.raise_for_status()

            # Update ETag from response
            new_etag = response.headers.get("etag")
            if new_etag:
                self._etag = new_etag
                logger.debug(f"Updated ETag: {new_etag}")

            # Get raw state bytes
            state_bytes = response.content

            # Cache the state
            self._cached_state = state_bytes

            logger.info(f"Fetched state successfully (size: {len(state_bytes)} bytes)")

            return state_bytes

        except httpx.HTTPError as e:
            logger.error(f"Failed to fetch state from CDN: {e}")
            # If we have cached state, return it to maintain availability
            if self._cached_state is not None:
                logger.warning("Returning cached state due to fetch error")
                return self._cached_state
            # No cache available, re-raise the error
            raise

    def get_cached_state(self) -> Optional[bytes]:
        """Get the cached state bytes if available.

        Returns:
            Cached state bytes or None if no cache
        """
        return self._cached_state
